File: internal/cleaner/cleaner.go
package cleaner

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	scannerPkg "github.com/Drilmo/git-secret-scanner/internal/scanner"
)

// CleanOptions holds cleaning options
type CleanOptions struct {
	Tool       string          // auto, filter-repo, bfg, filter-branch
	Source     string          // current, history, both
	FilePaths  map[string]bool // Files to clean (only these files will be modified for current files)
	DryRun     bool
	Force      bool
	NoBackup   bool
	OnProgress func(step, total int, message string)
}

// CleanResult holds cleaning results
type CleanResult struct {
	Tool           string
	Source         string
	SecretsRemoved int
	PatternsUsed   int
	FilesModified  int // Number of current files modified
	Success        bool
	Message        string
	BackupBranch   string
	DryRun         bool
	PreviewSecrets []string // First few secrets (masked) for preview
}

// Cleaner performs git history cleaning
type Cleaner struct{}

// New creates a new Cleaner
func New() *Cleaner {
	return &Cleaner{}
}

// HasFilterRepo checks if git-filter-repo is installed
func HasFilterRepo() bool {
	cmd := exec.Command("git", "filter-repo", "--version")
	return cmd.Run() == nil
}

// HasBFG checks if BFG is installed
func HasBFG() bool {
	cmd := exec.Command("bfg", "--version")
	if cmd.Run() == nil {
		return true
	}
	// Try java -jar bfg.jar
	cmd = exec.Command("java", "-jar", "bfg.jar", "--version")
	return cmd.Run() == nil
}

// GetAvailableTools returns list of available cleaning tools
func GetAvailableTools() map[string]bool {
	return map[string]bool{
		"filter-repo":   HasFilterRepo(),
		"bfg":           HasBFG(),
		"filter-branch": true, // Always available
	}
}

// Clean performs the cleaning operation
func (c *Cleaner) Clean(repoPath string, secrets []string, opts CleanOptions) (*CleanResult, error) {
	if len(secrets) == 0 {
		return &CleanResult{
			Success: true,
			Message: "No secrets to clean",
			DryRun:  opts.DryRun,
		}, nil
	}

	// Default source to "both"
	source := opts.Source
	if source == "" {
		source = "both"
	}

	// Select tool for history cleaning
	tool := opts.Tool
	if tool == "" || tool == "auto" {
		tool = selectBestTool()
	}

	// Group secrets into patterns
	patterns := groupSecretsIntoPatterns(secrets)

	// For dry run, prepare preview and return early
	if opts.DryRun {
		preview := make([]string, 0, min(10, len(secrets)))
		for i, s := range secrets {
			if i >= 10 {
				break
			}
			preview = append(preview, maskSecret(s))
		}

		var msg string
		switch source {
		case "current":
			msg = fmt.Sprintf("[DRY-RUN] Would clean %d secrets in current files only", len(secrets))
		case "history":
			msg = fmt.Sprintf("[DRY-RUN] Would clean %d secrets in git history using %s", len(secrets), tool)
		default:
			msg = fmt.Sprintf("[DRY-RUN] Would clean %d secrets in current files + git history using %s", len(secrets), tool)
		}

		return &CleanResult{
			Tool:           tool,
			Source:         source,
			SecretsRemoved: len(secrets),
			PatternsUsed:   len(patterns),
			Success:        true,
			Message:        msg,
			DryRun:         true,
			PreviewSecrets: preview,
		}, nil
	}

	// Create backup unless disabled (only for history cleaning)
	var backupBranch string
	if !opts.NoBackup && (source == "history" || source == "both") {
		backupBranch = fmt.Sprintf("backup-before-clean-%d", os.Getpid())
		cmd := exec.Command("git", "branch", backupBranch)
		cmd.Dir = repoPath
		cmd.Run()
	}

	var result *CleanResult
	var err error
	var filesModified int

	// Clean current files if needed
	if source == "current" || source == "both" {
		if opts.OnProgress != nil {
			opts.OnProgress(1, 3, "Cleaning current files...")
		}
		filesModified, err = c.cleanCurrentFiles(repoPath, secrets, opts.FilePaths)
		if err != nil {
			return &CleanResult{
				Success: false,
				Source:  source,
				Message: fmt.Sprintf("Failed to clean current files: %v", err),
			}, nil
		}
	}

	// Clean git history if needed
	if source == "history" || source == "both" {
		if opts.OnProgress != nil {
			step := 1
			if source == "both" {
				step = 2
			}
			opts.OnProgress(step, 3, fmt.Sprintf("Cleaning git history using %s with %d patterns", tool, len(patterns)))
		}

		switch tool {
		case "filter-repo":
			result, err = c.cleanWithFilterRepo(repoPath, patterns, opts)
		case "bfg":
			result, err = c.cleanWithBFG(repoPath, secrets, opts)
		default:
			result, err = c.cleanWithFilterBranch(repoPath, patterns, opts)
		}

		if err != nil {
			return nil, err
		}

		// Run git gc after history rewrite
		if result.Success {
			if opts.OnProgress != nil {
				opts.OnProgress(3, 3, "Running git gc...")
			}
			cmd := exec.Command("git", "reflog", "expire", "--expire=now", "--all")
			cmd.Dir = repoPath
			cmd.Run()

			cmd = exec.Command("git", "gc", "--prune=now", "--aggressive")
			cmd.Dir = repoPath
			cmd.Run()
		}
	} else {
		// Current files only - create simple success result
		result = &CleanResult{
			Success: true,
			Message: fmt.Sprintf("Successfully cleaned %d files", filesModified),
		}
	}

	result.Tool = tool
	result.Source = source
	result.SecretsRemoved = len(secrets)
	result.PatternsUsed = len(patterns)
	result.FilesModified = filesModified
	result.BackupBranch = backupBranch
	result.DryRun = false

	// Update message based on source
	if result.Success {
		switch source {
		case "current":
			result.Message = fmt.Sprintf("Successfully cleaned %d secrets in %d files (current files only)", len(secrets), filesModified)
		case "history":
			result.Message = fmt.Sprintf("Successfully cleaned %d secrets in git history using %s", len(secrets), tool)
		default:
			result.Message = fmt.Sprintf("Successfully cleaned %d secrets in %d files + git history using %s", len(secrets), filesModified, tool)
		}
	}

	return result, nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maskSecret(value string) string {
	if len(value) <= 4 {
		return "****"
	}
	maskLen := len(value) - 4
	if maskLen > 16 {
		maskLen = 16
	}
	return value[:2] + strings.Repeat("*", maskLen) + value[len(value)-2:]
}

func selectBestTool() string {
	if HasFilterRepo() {
		return "filter-repo"
	}
	if HasBFG() {
		return "bfg"
	}
	return "filter-branch"
}

// Group secrets into regex patterns (max 100 per pattern)
func groupSecretsIntoPatterns(secrets []string) []string {
	// Sort by length (longest first)
	sorted := make([]string, len(secrets))
	copy(sorted, secrets)
	sort.Slice(sorted, func(i, j int) bool {
		return len(sorted[i]) > len(sorted[j])
	})

	var patterns []string
	batchSize := 100

	for i := 0; i < len(sorted); i += batchSize {
		end := i + batchSize
		if end > len(sorted) {
			end = len(sorted)
		}

		batch := sorted[i:end]
		escaped := make([]string, len(batch))
		for j, s := range batch {
			escaped[j] = regexp.QuoteMeta(s)
		}

		pattern := "(" + strings.Join(escaped, "|") + ")"
		patterns = append(patterns, pattern)
	}

	return patterns
}

// cleanCurrentFiles replaces secrets in current files without rewriting git history
// Only files listed in allowedFiles will be modified (if nil, no files are modified)
func (c *Cleaner) cleanCurrentFiles(repoPath string, secrets []string, allowedFiles map[string]bool) (int, error) {
	filesModified := 0

	// If no allowed files specified, don't modify anything
	if allowedFiles == nil || len(allowedFiles) == 0 {
		return 0, nil
	}

	// Only process files that are in the allowed list
	for filePath := range allowedFiles {
		// Build full path
		fullPath := filepath.Join(repoPath, filePath)

		// Check if file exists
		info, err := os.Stat(fullPath)
		if err != nil {
			continue // Skip files that don't exist
		}

		// Skip directories
		if info.IsDir() {
			continue
		}

		// Skip large files (> 1MB)
		if info.Size() > 1024*1024 {
			continue
		}

		// Read file content
		content, err := os.ReadFile(fullPath)
		if err != nil {
			continue
		}

		// Check if file contains any secrets and replace them
		modified := false
		contentStr := string(content)
		for _, secret := range secrets {
			if strings.Contains(contentStr, secret) {
				contentStr = strings.ReplaceAll(contentStr, secret, "***REMOVED***")
				modified = true
			}
		}

		// Write back if modified
		if modified {
			if err := os.WriteFile(fullPath, []byte(contentStr), info.Mode()); err != nil {
				continue
			}
			filesModified++
		}
	}

	return filesModified, nil
}

func (c *Cleaner) cleanWithFilterRepo(repoPath string, patterns []string, opts CleanOptions) (*CleanResult, error) {
	// Create replacements file
	replacementsFile := fmt.Sprintf("/tmp/replacements-%d.txt", os.Getpid())
	f, err := os.Create(replacementsFile)
	if err != nil {
		return nil, err
	}

	for _, pattern := range patterns {
		f.WriteString(fmt.Sprintf("regex:%s==>***REMOVED***\n", pattern))
	}
	f.Close()
	defer os.Remove(replacementsFile)

	args := []string{"filter-repo", "--replace-text", replacementsFile}
	if opts.Force {
		args = append(args, "--force")
	}

	cmd := exec.Command("git", args...)
	cmd.Dir = repoPath
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return &CleanResult{
			Success: false,
			Message: fmt.Sprintf("git-filter-repo failed: %v", err),
		}, nil
	}

	return &CleanResult{
		Success: true,
		Message: "Successfully cleaned with git-filter-repo",
	}, nil
}

func (c *Cleaner) cleanWithBFG(repoPath string, secrets []string, opts CleanOptions) (*CleanResult, error) {
	// Create replacements file
	replacementsFile := fmt.Sprintf("/tmp/bfg-replacements-%d.txt", os.Getpid())
	f, err := os.Create(replacementsFile)
	if err != nil {
		return nil, err
	}

	for _, secret := range secrets {
		f.WriteString(secret + "\n")
	}
	f.Close()
	defer os.Remove(replacementsFile)

	cmd := exec.Command("bfg", "--replace-text", replacementsFile, repoPath)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return &CleanResult{
			Success: false,
			Message: fmt.Sprintf("BFG failed: %v", err),
		}, nil
	}

	return &CleanResult{
		Success: true,
		Message: "Successfully cleaned with BFG",
	}, nil
}

func (c *Cleaner) cleanWithFilterBranch(repoPath string, patterns []string, opts CleanOptions) (*CleanResult, error) {
	// Build sed command
	sedParts := make([]string, len(patterns))
	for i, pattern := range patterns {
		sedParts[i] = fmt.Sprintf("s/%s/***REMOVED***/g", pattern)
	}
	sedCommand := strings.Join(sedParts, "; ")

	filterCommand := fmt.Sprintf(`git ls-files -z | xargs -0 sed -i '' '%s' 2>/dev/null || true`, sedCommand)

	cmd := exec.Command("git", "filter-branch", "-f", "--tree-filter", filterCommand, "--", "--all")
	cmd.Dir = repoPath
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return &CleanResult{
			Success: false,
			Message: fmt.Sprintf("git-filter-branch failed: %v", err),
		}, nil
	}

	return &CleanResult{
		Success: true,
		Message: "Successfully cleaned with git-filter-branch (slow method)",
	}, nil
}

// LoadSecretsResult holds secrets and detected source
type LoadSecretsResult struct {
	Secrets   []string
	FilePaths []string          // List of file paths containing secrets
	FileMap   map[string]bool   // Map of file paths for quick lookup
	Source    string            // "current", "history", or "both"
}

// LoadSecretsFromJSONL loads secrets from a JSONL file and detects source
func LoadSecretsFromJSONL(path string) (*LoadSecretsResult, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	values := make(map[string]bool)
	filePaths := make(map[string]bool)
	hasCurrent := false
	hasHistory := false
	fileScanner := bufio.NewScanner(file)

	for fileScanner.Scan() {
		var entry scannerPkg.StreamEntry
		if err := json.Unmarshal(fileScanner.Bytes(), &entry); err != nil {
			continue
		}
		if entry.Value != "" && !strings.Contains(entry.Value, "REMOVED") {
			values[entry.Value] = true

			// Track file path for current files
			if entry.File != "" {
				filePaths[entry.File] = true
			}

			// Detect source from commit field
			if entry.Commit == "current" {
				hasCurrent = true
			} else if entry.Commit != "" {
				hasHistory = true
			}
		}
	}

	secrets := make([]string, 0, len(values))
	for v := range values {
		secrets = append(secrets, v)
	}

	paths := make([]string, 0, len(filePaths))
	for p := range filePaths {
		paths = append(paths, p)
	}

	// Determine source
	source := "both"
	if hasCurrent && !hasHistory {
		source = "current"
	} else if hasHistory && !hasCurrent {
		source = "history"
	}

	return &LoadSecretsResult{
		Secrets:   secrets,
		FilePaths: paths,
		FileMap:   filePaths,
		Source:    source,
	}, nil
}

// LoadSecretsFromJSON loads secrets from a JSON scan result file and detects source
func LoadSecretsFromJSON(path string) (*LoadSecretsResult, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var result scannerPkg.ScanResult
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	// Detect source from commits in history and collect file paths
	hasCurrent := false
	hasHistory := false
	filePaths := make(map[string]bool)

	for _, secret := range result.Secrets {
		// Track file path
		if secret.File != "" {
			filePaths[secret.File] = true
		}

		for _, h := range secret.History {
			for _, commit := range h.Commits {
				if commit == "current" {
					hasCurrent = true
				} else if commit != "" {
					hasHistory = true
				}
			}
		}
	}

	paths := make([]string, 0, len(filePaths))
	for p := range filePaths {
		paths = append(paths, p)
	}

	// Determine source
	source := "both"
	if hasCurrent && !hasHistory {
		source = "current"
	} else if hasHistory && !hasCurrent {
		source = "history"
	}

	return &LoadSecretsResult{
		Secrets:   scannerPkg.GetAllValues(&result),
		FilePaths: paths,
		FileMap:   filePaths,
		Source:    source,
	}, nil
}

File: python/gitsecret/cleaner.py
import dataclasses
import json
import os
import re
import subprocess
import tempfile
from pathlib import Path
from typing import Callable, Dict, List, Optional


@dataclasses.dataclass
class CleanOptions:
    """Options for cleaning secrets from git history."""
    tool: str = "auto"
    source: str = "both"
    file_paths: Optional[Dict[str, bool]] = None
    dry_run: bool = False
    force: bool = False
    no_backup: bool = False
    on_progress: Optional[Callable[[str], None]] = None


@dataclasses.dataclass
class CleanResult:
    """Result of a cleaning operation."""
    tool: str = ""
    source: str = ""
    secrets_removed: int = 0
    patterns_used: int = 0
    files_modified: int = 0
    success: bool = False
    message: str = ""
    backup_branch: str = ""
    dry_run: bool = False
    preview_secrets: Optional[List[str]] = None


@dataclasses.dataclass
class LoadSecretsResult:
    """Result of loading secrets from a file."""
    secrets: List[str]
    file_paths: List[str]
    file_map: Dict[str, int]
    source: str


def has_filter_repo() -> bool:
    """Check if git-filter-repo is available."""
    try:
        subprocess.run(
            ["git", "filter-repo", "--version"],
            capture_output=True,
            timeout=5
        )
        return True
    except (subprocess.SubprocessError, FileNotFoundError):
        return False


def has_bfg() -> bool:
    """Check if BFG is available."""
    try:
        subprocess.run(
            ["bfg", "--version"],
            capture_output=True,
            timeout=5
        )
        return True
    except (subprocess.SubprocessError, FileNotFoundError):
        pass

    try:
        subprocess.run(
            ["java", "-jar", "bfg.jar", "--version"],
            capture_output=True,
            timeout=5
        )
        return True
    except (subprocess.SubprocessError, FileNotFoundError):
        return False


def get_available_tools() -> Dict[str, bool]:
    """Get dict of available tools."""
    return {
        "filter-repo": has_filter_repo(),
        "bfg": has_bfg(),
        "filter-branch": True,
    }


def mask_secret(value: str) -> str:
    """Mask a secret value: show first 2 and last 2 chars, replace middle with asterisks."""
    if len(value) <= 4:
        return "****"
    mask_len = min(len(value) - 4, 16)
    return value[:2] + ("*" * mask_len) + value[-2:]


class Cleaner:
    """Cleans secrets from git repositories."""

    def clean(
        self,
        repo_path: str,
        secrets: List[str],
        opts: CleanOptions,
    ) -> CleanResult:
        """Clean secrets from git history and/or current files."""
        result = CleanResult()

        if not secrets:
            result.success = True
            result.message = "No secrets to clean"
            return result

        if not opts.source:
            opts.source = "both"

        result.source = opts.source
        result.dry_run = opts.dry_run

        if opts.tool == "auto":
            tool = self._select_best_tool()
        else:
            tool = opts.tool

        result.tool = tool

        patterns = self._group_secrets_into_patterns(secrets)
        result.patterns_used = len(patterns)

        if opts.dry_run:
            preview = [mask_secret(s) for s in secrets[:10]]
            result.preview_secrets = preview
            result.success = True
            result.message = f"Dry run: would remove {len(secrets)} secrets using {tool}"
            return result

        if not opts.no_backup:
            pid = os.getpid()
            backup_name = f"backup-before-clean-{pid}"
            try:
                subprocess.run(
                    ["git", "branch", backup_name],
                    cwd=repo_path,
                    capture_output=True,
                    check=True,
                    timeout=30,
                )
                result.backup_branch = backup_name
            except subprocess.SubprocessError as e:
                result.success = False
                result.message = f"Failed to create backup branch: {e}"
                return result

        if opts.source in ("current", "both"):
            allowed_files = opts.file_paths if opts.file_paths else {}
            modified = self._clean_current_files(repo_path, secrets, allowed_files)
            result.files_modified = modified

        if opts.source in ("history", "both"):
            try:
                if tool == "filter-repo":
                    self._clean_with_filter_repo(repo_path, patterns, opts)
                elif tool == "bfg":
                    self._clean_with_bfg(repo_path, secrets, opts)
                else:
                    self._clean_with_filter_branch(repo_path, patterns, opts)
            except subprocess.SubprocessError as e:
                result.success = False
                result.message = f"Failed to clean history with {tool}: {e}"
                return result

        try:
            subprocess.run(
                ["git", "reflog", "expire", "--expire=now", "--all"],
                cwd=repo_path,
                capture_output=True,
                timeout=60,
            )
            subprocess.run(
                ["git", "gc", "--prune=now", "--aggressive"],
                cwd=repo_path,
                capture_output=True,
                timeout=120,
            )
        except subprocess.SubprocessError as e:
            result.success = False
            result.message = f"Failed to clean up git: {e}"
            return result

        result.success = True
        result.secrets_removed = len(secrets)
        result.message = f"Successfully cleaned {len(secrets)} secrets from {opts.source}"

        return result

    def _clean_current_files(
        self,
        repo_path: str,
        secrets: List[str],
        allowed_files: Dict[str, bool],
    ) -> int:
        """Clean secrets from current files."""
        modified_count = 0

        if not allowed_files:
            try:
                result = subprocess.run(
                    ["git", "ls-files"],
                    cwd=repo_path,
                    capture_output=True,
                    text=True,
                    timeout=30,
                )
                tracked_files = result.stdout.strip().split("\n")
                allowed_files = {f: True for f in tracked_files if f}
            except subprocess.SubprocessError:
                return 0

        for file_path in allowed_files:
            full_path = os.path.join(repo_path, file_path)

            if not os.path.isfile(full_path):
                continue

            try:
                with open(full_path, "r", encoding="utf-8", errors="ignore") as f:
                    content = f.read()

                original_content = content

                for secret in secrets:
                    content = content.replace(secret, "***REMOVED***")

                if content != original_content:
                    with open(full_path, "w", encoding="utf-8") as f:
                        f.write(content)
                    modified_count += 1
            except (OSError, IOError):
                continue

        return modified_count

    def _clean_with_filter_repo(
        self,
        repo_path: str,
        patterns: List[str],
        opts: CleanOptions,
    ) -> None:
        """Clean using git-filter-repo."""
        with tempfile.NamedTemporaryFile(
            mode="w",
            suffix=".txt",
            delete=False,
            dir=repo_path,
        ) as f:
            for pattern in patterns:
                f.write(f"regex:{pattern}===>***REMOVED***\n")
            temp_file = f.name

        try:
            cmd = ["git", "filter-repo", "--replace-text", temp_file]
            if opts.force:
                cmd.append("--force")

            subprocess.run(
                cmd,
                cwd=repo_path,
                check=True,
                timeout=600,
            )
        finally:
            try:
                os.unlink(temp_file)
            except OSError:
                pass

    def _clean_with_bfg(
        self,
        repo_path: str,
        secrets: List[str],
        opts: CleanOptions,
    ) -> None:
        """Clean using BFG."""
        with tempfile.NamedTemporaryFile(
            mode="w",
            suffix=".txt",
            delete=False,
            dir=repo_path,
        ) as f:
            for secret in secrets:
                f.write(secret + "\n")
            temp_file = f.name

        try:
            cmd = ["bfg", "--replace-text", temp_file, repo_path]

            subprocess.run(
                cmd,
                cwd=repo_path,
                check=True,
                timeout=600,
            )
        except FileNotFoundError:
            cmd = ["java", "-jar", "bfg.jar", "--replace-text", temp_file, repo_path]
            subprocess.run(
                cmd,
                cwd=repo_path,
                check=True,
                timeout=600,
            )
        finally:
            try:
                os.unlink(temp_file)
            except OSError:
                pass

    def _clean_with_filter_branch(
        self,
        repo_path: str,
        patterns: List[str],
        opts: CleanOptions,
    ) -> None:
        """Clean using git-filter-branch."""
        sed_exprs = []
        for pattern in patterns:
            escaped_pattern = pattern.replace("~", "\\~")
            sed_exprs.append(f"s~{escaped_pattern}~***REMOVED***~g")

        sed_script = ";".join(sed_exprs)

        with tempfile.NamedTemporaryFile(
            mode="w",
            suffix=".sed",
            delete=False,
            dir=repo_path,
        ) as f:
            f.write(sed_script)
            sed_file = f.name

        try:
            tree_filter = f"find . -type f ! -path './.git*' -exec sed -i -f {sed_file} {{}} + 2>/dev/null || true"

            cmd = [
                "git",
                "filter-branch",
                "-f",
                "--tree-filter",
                tree_filter,
                "--",
                "--all",
            ]

            subprocess.run(
                cmd,
                cwd=repo_path,
                check=True,
                timeout=600,
            )
        finally:
            try:
                os.unlink(sed_file)
            except OSError:
                pass

    @staticmethod
    def load_secrets_from_jsonl(path: str) -> LoadSecretsResult:
        """Load secrets from JSONL file."""
        secrets = []
        file_paths = []
        file_map: Dict[str, int] = {}
        source_set = set()

        with open(path, "r") as f:
            for line in f:
                if not line.strip():
                    continue

                item = json.loads(line)

                if "value" in item:
                    secrets.append(item["value"])

                if "file" in item:
                    fp = item["file"]
                    if fp not in file_paths:
                        file_paths.append(fp)
                    file_map[fp] = file_map.get(fp, 0) + 1

                if "commit" in item and item["commit"]:
                    source_set.add("history")
                else:
                    source_set.add("current")

        if "history" in source_set and "current" in source_set:
            source_str = "both"
        elif "history" in source_set:
            source_str = "history"
        else:
            source_str = "current"

        secrets = list(set(secrets))

        return LoadSecretsResult(
            secrets=secrets,
            file_paths=file_paths,
            file_map=file_map,
            source=source_str,
        )

    @staticmethod
    def load_secrets_from_json(path: str) -> LoadSecretsResult:
        """Load secrets from JSON file."""
        with open(path, "r") as f:
            data = json.load(f)

        secrets = []
        file_paths = []
        file_map: Dict[str, int] = {}
        source_set = set()

        if "results" in data:
            for result in data["results"]:
                if "value" in result:
                    secrets.append(result["value"])

                if "file" in result:
                    fp = result["file"]
                    if fp not in file_paths:
                        file_paths.append(fp)
                    file_map[fp] = file_map.get(fp, 0) + 1

                if "commit" in result and result["commit"]:
                    source_set.add("history")
                else:
                    source_set.add("current")

        if "history" in source_set and "current" in source_set:
            source_str = "both"
        elif "history" in source_set:
            source_str = "history"
        else:
            source_str = "current"

        secrets = list(set(secrets))
        secrets.sort(key=len, reverse=True)

        return LoadSecretsResult(
            secrets=secrets,
            file_paths=file_paths,
            file_map=file_map,
            source=source_str,
        )

    def _group_secrets_into_patterns(self, secrets: List[str]) -> List[str]:
        """Group secrets into regex patterns (batches of 100)."""
        patterns = []
        batch = []

        for secret in secrets:
            batch.append(re.escape(secret))

            if len(batch) >= 100:
                patterns.append("|".join(batch))
                batch = []

        if batch:
            patterns.append("|".join(batch))

        return patterns

    def _select_best_tool(self) -> str:
        """Select the best available tool."""
        if has_filter_repo():
            return "filter-repo"
        elif has_bfg():
            return "bfg"
        else:
            return "filter-branch"


# Module-level convenience functions (delegates to Cleaner static methods)
def load_secrets_from_jsonl(path: str) -> LoadSecretsResult:
    """Load secrets from a JSONL file and detect source."""
    return Cleaner.load_secrets_from_jsonl(path)


def load_secrets_from_json(path: str) -> LoadSecretsResult:
    """Load secrets from a JSON scan result file and detect source."""
    return Cleaner.load_secrets_from_json(path)
